### PR TITLE
 Remove OOT photons from 2018 heavy-ion (AA) era 

### PR DESCRIPTION
**Pull request by mandrenguyen**
OOT photons are not needed for reconstruction of PbPb events. 
This PR modifies the era "pp_on_AA_2018", which is intended for PbPb processing this year.
It was tested with wf 158.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
please test

**Comment by mandrenguyen** [comment]
I am unable to reproduce this error by running the failed addOn test.  From the logfile the crash does not appear to be obviously related to this pull request.  Should we give it another 'please test'?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
the crash was  in FastSimProducer:fastSimProducer related to 
``` #8  0x00007fe47478e419 in fastsim::TrackerSimHitProducer::interact```

**Comment by slava77** [comment]
@mandrenguyen 
I'm getting a crash in 158.0

```
   [1] Prefetching for module TopMonitor/'AK8PFJet140'
   [2] Prefetching for module VersionedGsfElectronIdProducer/'egmGsfElectronIDsForDQM'
   [3] Calling method for module ReducedEGProducer/'reducedEgamma'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<reco::Photon>
Looking for module label: ootPhotons
```

a change is missing in reducedEgamma_cfi.py to set ```ootPhotons```  to ```""``` for the HI era

**Comment by mandrenguyen** [comment]
@slava77 I guess I'm confused about how this is consistent with your previous comment:
"I think that this change is not necessary.
Simply test for a value of ootPhotonT_.isUnintialized() in places where this doOotPhotons_ was introduced"
Not sure why I didn't see the crash.  I will add the flag.

**Comment by slava77** [comment]
On 6/4/18 9:06 AM, Matthew Nguyen wrote:
> @slava77 <https://github.com/slava77> I guess I'm confused about how
> this is consistent with your previous comment:
> "I think that this change is not necessary.
> Simply test for a value of ootPhotonT_.isUnintialized() in places where
> this doOotPhotons_ was introduced"
> Not sure why I didn't see the crash. I will add the flag.
> 
> —

my comment was to the case where you introduced an additional data
member which detects the empty string in the inputTag. That extra data
member is not needed.
The empty string still has to be set in the configuration.

**Comment by slava77** [comment]
uhm, a check for an empty input is actually also needed when ```ootPhotonT_```

(remove ```ootPhotonT_``` from the initializer list)
```
auto const& aTag = config.getParameter<edm::InputTag>("ootPhotons");
if (not aTag.label().empty()) ootPhotonT_ = consumes<reco::PhotonCollection>(aTag);
```

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test workflow 158.0

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 1]
```diff
@@ -558,9 +571,11 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
       eeRecHits->push_back(rechit);
     }
   }
- 
-  theEvent.put(std::move(ebRecHits),outEBRecHits_);
-  theEvent.put(std::move(eeRecHits),outEERecHits_);
+
+  if(!ootPhotonT_.isUninitialized()) {
+    theEvent.put(std::move(ebRecHits),outEBRecHits_);
+    theEvent.put(std::move(eeRecHits),outEERecHits_);
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: why are these hidden/removed? these have nothing to do with OOT.

**Comment by slava77** [comment]
I see a crash after a few events in 158.0 (it's event 8 in my case, but I made the GEN-SIM in multithreaded, which may not be exactly reproducible).

```
   [5] Prefetching for module VersionedGsfElectronIdProducer/'egmGsfElectronIDs'
   [6] Calling method for module ElectronHEEPIDValueMapProducer/'heepIDVarValueMaps'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::SortedCollection<EcalRecHit,edm::StrictWeakOrdering<EcalRecHit> >
Looking for module label: reducedEgamma
Looking for productInstanceName: reducedEERecHits

```
this looks related to https://github.com/cms-sw/cmssw/pull/23371#pullrequestreview-125952200

**Comment by slava77** [comment]
@cmsbuild please test workflow 158.0

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/23371	56a39ea
- code changes are in line with the PR description; the changes affect only Run2_2018_pp_on_AA era
- jenkins tests pass
- local tests on 200 events of wf 158.0 (using Run2_2018_pp_on_AA) show expected behavior: OOT-related modules are gone from the execution paths. Savings on this sample are rather minimal, adding up to about 30 ms/ev out of about 9 s/ev in reco modules total.

@mandrenguyen was the OOT a significant contributor in some of your tests? It seems to me that the strategy to file away small contributions from pp like this is particularly practical.

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Comment by mandrenguyen** [comment]
@slava77 I tested the timing together with other customizations that are included in #23404, as well as in a forthcoming PR.  
The full set I tested is listed here:
https://twiki.cern.ch/twiki/pub/CMS/HiEgamma2018/egamma_customisations.txt
I would have to dig up the total savings, which are not negligible. 

We decided to split this into several PRs.  Unfortunately, this one doesn't seem to be save much.

**Pull request merged**